### ISSUE TITLE
SABOTAGE attribute not shown in home-assistant (IP presence sensor)

### ISSUE THREAD

**Issue by akloeckner** [danielperna84/pyhomematic#108]
Hi there, I'm running pyhomematic as part of home-assistant and I can't seem to find a way to access the "SABOTAGE" attribute of my IP-Presence-Detector. I have tried to check the code for this and there seems to be provisions for filling the ATTRIBUTENODE here: 

https://github.com/danielperna84/pyhomematic/blob/68b5e3ef67f9231047d83afedb80e0a2c49d0777/pyhomematic/devicetypes/sensors.py#L300 

There also seems to be provisions to do something with this ATTRIBUTENODE here:

https://github.com/home-assistant/home-assistant/blob/75a39352ffb8c5e7b24fe10a9a66d0839224d3e9/homeassistant/components/homematic/__init__.py#L824

My home-assistant log shows this:

```
hass[6380]: 2018-01-06 00:32:32 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
```

So, something arives at the system. But I am not able to access this attribute within home-assistant, by doing e.g. this:

```
{{states.binary_sensor.flur_sensor_presence_detection_state.attributes.SABOTAGE}}
```

I might just be to new too this, then please someone tipp me off! Or it is somehow not propagated. Unfortunately, I am not so current in python, pyhomematic nor home-assistant to find the problem myself...

Thanks! Andreas

**Comment by danielperna84** [comment]
I think that's actually a bug. I have shutter-contacts where the sabotage is working correctly. Unfortunately I don't have the device to verify and fix it. So we'll have to wait until someone comes along who's capable of doing that.

**Comment by akloeckner** [comment]
I now checked that. Thanks for the hint. The two sensors show the following (without sabotage triggered):
```
{
  "interface": "ip",
  "rssi": -25,
  "friendly_name": "Flur Sensor PRESENCE_DETECTION_STATE",
  "id": "000C170999****",
  "device_class": "motion"
}

{
  "interface": "ip",
  "rssi": -25,
  "friendly_name": "Flur Sensor ILLUMINATION",
  "id": "000C170999****",
  "unit_of_measurement": "lux"
}
```

Then I opened the sensor. The entities read:

```
{
  "interface": "ip",
  "rssi": -25,
  "friendly_name": "Flur Sensor PRESENCE_DETECTION_STATE",
  "id": "000C170999****",
  "device_class": "motion"
}

{
  "interface": "ip",
  "rssi": -25,
  "friendly_name": "Flur Sensor ILLUMINATION",
  "id": "000C170999****",
  "unit_of_measurement": "lux"
}
```

After re-attaching the sensor I pulled the following log:

```
osmc@ak-test:~$ sudo journalctl  -u homeassistant@homeassistant|grep SABOTAGE
Jan 07 20:50:59 ak-test hass[306]: 2018-01-07 20:50:59 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 20:53:08 ak-test hass[306]: 2018-01-07 20:53:08 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 20:56:07 ak-test hass[306]: 2018-01-07 20:56:07 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 20:58:52 ak-test hass[306]: 2018-01-07 20:58:52 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 20:58:54 ak-test hass[306]: 2018-01-07 20:58:54 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 20:58:55 ak-test hass[306]: 2018-01-07 20:58:55 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=True
Jan 07 20:58:57 ak-test hass[306]: 2018-01-07 20:58:57 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 20:59:59 ak-test hass[306]: 2018-01-07 20:59:59 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 21:01:22 ak-test hass[306]: 2018-01-07 21:01:22 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 21:02:16 ak-test hass[306]: 2018-01-07 21:02:16 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Jan 07 21:02:17 ak-test hass[306]: 2018-01-07 21:02:17 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=True
Jan 07 21:02:19 ak-test hass[306]: 2018-01-07 21:02:19 INFO (Thread-22) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999****:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
```

Note, how the SABOTAGE key get's True just for one instant and then changes back. Both when "sabotaging" and when "fixing" the sensor.

I also noticed the customised "lux" unit_of_measurement. Commenting that out in the yaml doesn't change the output as listed above...

Ok... let's see if someone with this device and with some more experience in pyhomematic finds it enough bothering to fix it. ;-)

**Comment by akloeckner** [comment]
I just thad a similar problem, which might help pinpointing the bug:

Today, I noted that the LOW_BAT attribute was also not showing in HomeAssistant. I found that startling, because I knew I had it showing correctly before. So, I checked in the logs and there are lots of messages like this: 
```
 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=LOW_BAT, value=True
```
But that did not arrive in the HA states. Which reminded me of the SABOTAGE issue.

Now, interestingly, while playing around, I noticed that the `LOW_BAT=True` translated well to a `battery=Low` attribute when restarting HA!

And the same works for the sabotage attribute: I sabotaged the sensor (plugged it off the wall), restarted HA and the sensor now shows `sabotage=Yes`. But it did not update to `sabotage=No`, when I re-mounted the sensor to the wall. 

Summary: The attributes LOW_BAT and SABOTAGE seem to be imported into HA only on creating the entity.

Maybe that rings a bell with you, @danielperna84 ? I tried to work through the code, but I still don't understand enough to have the slightest idea, where to check...

**Comment by danielperna84** [comment]
There was a breaking change in version 0.89 (last one in [this](https://www.home-assistant.io/blog/2019/03/06/release-89/#breaking-changes) list) of Home Assistant, which has the potential to fix this issue. Have you updated to the latest release yet and could have a look if the change actually does solve the problem?

**Comment by akloeckner** [comment]
Thanks for the hint! I have now updated and unfortunately this issue is not solved.

**Comment by danielperna84** [comment]
Earlier you have posted:
`{{states.binary_sensor.flur_sensor_presence_detection_state.attributes.SABOTAGE}}`

The attributes are lower case though. So the correct variant would be:
`{{states.binary_sensor.flur_sensor_presence_detection_state.attributes.sabotage}}`

Besides that I don't know what else could be causing this problem.

**Comment by akloeckner** [comment]
I tried lowercase just to be sure. You're right in that I probably made this mistake and could therefore not "access this attribute". I can access the lowercase attribute (as verified just now). However, I kept actually looking at the entity in my HA GUI. Unfortunately, there is still no change, when I sabotage my sensor.

Again just to be sure, I also waited for the sensor to change its primary state before remounting the sensor. (The attribute might be updated only when the primary sensor updated.) But that also didn't change the attribute...

The logs still show something arriving at the system. Also note the `ERROR_CODE` which is not reflected in the HA GUI:
```
Mar 22 01:29:18 ak-test hass[26692]: 2019-03-22 01:29:18 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=SABOTAGE, value=True
Mar 22 01:29:18 ak-test hass[26692]: 2019-03-22 01:29:18 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=LOW_BAT, value=False
Mar 22 01:29:18 ak-test hass[26692]: 2019-03-22 01:29:18 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=ERROR_CODE, value=1
Mar 22 01:29:18 ak-test hass[26692]: 2019-03-22 01:29:18 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=PRESENCE_DETECTION_STATE, value=True
[...]
Mar 22 01:29:19 ak-test hass[26692]: 2019-03-22 01:29:19 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Mar 22 01:29:19 ak-test hass[26692]: 2019-03-22 01:29:19 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=LOW_BAT, value=False
Mar 22 01:29:19 ak-test hass[26692]: 2019-03-22 01:29:19 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=ERROR_CODE, value=0
Mar 22 01:29:19 ak-test hass[26692]: 2019-03-22 01:29:19 INFO (Thread-17) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=PRESENCE_DETECTION_STATE, value=True
```

I only have these attributes in the GUI:
```
id 000C170999F54D
sabotage No
interface ip
battery High
rssi device -32
voltage 2.7
```

**Comment by akloeckner** [comment]
TL;DR: The `SABOTAGE` attribute lives on channel `0`. No `_hm_event_callback` function is called for `channel=0`. This might be caused by them being assigned to `self._eventcallbacks`; instead of `self._hmchannels[channel]._eventcallbacks` as is the case for `channel>0`. 

So, I looked into this again. Let me explain my findings:

0.  I assume, `SABOTAGE` is the attribute we are looking for.
1. The attributes of the sensor are retrieved from the python object in the `device_state_attributes` function: https://github.com/home-assistant/home-assistant/blob/2871a650f69257fa6e8e4254527007f964e2f806/homeassistant/components/homematic/__init__.py#L752
2. To this end, the specific attribute must be listed in `HM_ATTRIBUTE_SUPPORT`, which it is: https://github.com/home-assistant/home-assistant/blob/2871a650f69257fa6e8e4254527007f964e2f806/homeassistant/components/homematic/__init__.py#L107
3. Also, the specific attribute must be present in `self._data`, which it is, due to the `_load_data_from_hm` function: https://github.com/home-assistant/home-assistant/blob/2871a650f69257fa6e8e4254527007f964e2f806/homeassistant/components/homematic/__init__.py#L848
4.  There the attribute is read using `getAttributeData`, if the attribute is listed in `ATTRIBUTENODE`, which it is: https://github.com/danielperna84/pyhomematic/blob/9d438bb6d4850d7833602ef4703f900c067ccfd6/pyhomematic/devicetypes/helper.py#L24
5. However, this function (`_load_data_from_hm`) is only run during initialization of the object.
6. In order to update the attribute during runtime, an event callback needs to be registered, in `_subscribe_homematic_events` :https://github.com/home-assistant/home-assistant/blob/2871a650f69257fa6e8e4254527007f964e2f806/homeassistant/components/homematic/__init__.py#L815
7. For this, we need an entry in `ATTRIBUTENODE` and in `self._data`. Both are present as described above. An event handler is therefore registered for the specific channel `0` here: https://github.com/home-assistant/home-assistant/blob/2871a650f69257fa6e8e4254527007f964e2f806/homeassistant/components/homematic/__init__.py#L844, which can also be verified in the logs:
```
Apr 22 19:48:13 ak-test hass[32639]: 2019-04-22 19:48:13 DEBUG (Thread-22) [homeassistant.components.homematic] Subscribe channel 0 from Flur Sensor ILLUMINATION
Apr 22 19:48:13 ak-test hass[32639]: 2019-04-22 19:48:13 DEBUG (Thread-5) [homeassistant.components.homematic] Subscribe channel 0 from Flur Sensor PRESENCE_DETECTION_STATE
Apr 22 19:48:13 ak-test hass[32639]: 2019-04-22 19:48:13 DEBUG (Thread-22) [homeassistant.components.homematic] Subscribe channel 1 from Flur Sensor ILLUMINATION
Apr 22 19:48:13 ak-test hass[32639]: 2019-04-22 19:48:13 DEBUG (Thread-5) [homeassistant.components.homematic] Subscribe channel 1 from Flur Sensor PRESENCE_DETECTION_STATE
```
8. Now, each event either on channel `0` or channel `1` should trigger a callback in the `event` function, here: https://github.com/danielperna84/pyhomematic/blob/9d438bb6d4850d7833602ef4703f900c067ccfd6/pyhomematic/devicetypes/generic.py#L54
9. This can nicely be tracked in the debug logs:
```
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=SABOTAGE, value=False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=CONFIG_PENDING, value=False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=DUTY_CYCLE, value=False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=LOW_BAT, value=False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=UNREACH, value=False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=RSSI_DEVICE, value=-52
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=OPERATING_VOLTAGE_STATUS, value=0
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=OPERATING_VOLTAGE, value=2.5
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=ERROR_CODE, value=0
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=PRESENCE_DETECTION_STATE, value=False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: on>>
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'PRESENCE_DETECTION_STATE' value: False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 2151.0>>
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'PRESENCE_DETECTION_STATE' value: False
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=PRESENCE_DETECTION_ACTIVE, value=True
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: off>>
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'PRESENCE_DETECTION_ACTIVE' value: True
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 2151.0>>
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'PRESENCE_DETECTION_ACTIVE' value: True
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=ILLUMINATION, value=1927.0
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: off>>
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'ILLUMINATION' value: 1927.0
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 2151.0>>
Apr 23 17:14:54 ak-test hass[9404]: 2019-04-23 17:14:54 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'ILLUMINATION' value: 1927.0
Apr 23 17:14:55 ak-test hass[9404]: 2019-04-23 17:14:55 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=ILLUMINATION_STATUS, value=0
Apr 23 17:14:55 ak-test hass[9404]: 2019-04-23 17:14:55 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: off>>
Apr 23 17:14:55 ak-test hass[9404]: 2019-04-23 17:14:55 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'ILLUMINATION_STATUS' value: 0
Apr 23 17:14:55 ak-test hass[9404]: 2019-04-23 17:14:55 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 1927.0>>
Apr 23 17:14:55 ak-test hass[9404]: 2019-04-23 17:14:55 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'ILLUMINATION_STATUS' value: 0
```
and for a sabotaged sensor:
```
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=SABOTAGE, value=True
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=CONFIG_PENDING, value=False
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=DUTY_CYCLE, value=False
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=LOW_BAT, value=False
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=UNREACH, value=False
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=RSSI_DEVICE, value=-54
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=OPERATING_VOLTAGE_STATUS, value=0
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=OPERATING_VOLTAGE, value=2.6
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:0, interface_id=homeassistant-ip, key=ERROR_CODE, value=1
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=PRESENCE_DETECTION_STATE, value=True
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: on>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'PRESENCE_DETECTION_STATE' value: True
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 1927.0>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'PRESENCE_DETECTION_STATE' value: True
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=PRESENCE_DETECTION_ACTIVE, value=True
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: on>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'PRESENCE_DETECTION_ACTIVE' value: True
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 1927.0>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'PRESENCE_DETECTION_ACTIVE' value: True
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=ILLUMINATION, value=1927.0
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: on>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'ILLUMINATION' value: 1927.0
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 1927.0>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'ILLUMINATION' value: 1927.0
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 INFO (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: address=000C170999F54D:1, interface_id=homeassistant-ip, key=ILLUMINATION_STATUS, value=0
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor PRESENCE_DETECTION_STATE: on>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor PRESENCE_DETECTION_STATE received event 'ILLUMINATION_STATUS' value: 0
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [pyhomematic.devicetypes.generic] HMGeneric.event: Using callback <bound method HMDevice._hm_event_callback of <Entity Flur Sensor ILLUMINATION: 1927.0>>
Apr 23 17:21:40 ak-test hass[9404]: 2019-04-23 17:21:40 DEBUG (Thread-19) [homeassistant.components.homematic] Flur Sensor ILLUMINATION received event 'ILLUMINATION_STATUS' value: 0
```

8. Now, I would assume callbacks to be called back, every time the channel `0` receives an event. However, only channel `1` seems to forward the events to the handlers. Nothing happens for channel `0`. Apparently, the callbacks are somehow different. So, I followed the callback definition...
9. The callbacks are registered via the  `setEventCallback` function here: https://github.com/danielperna84/pyhomematic/blob/9d438bb6d4850d7833602ef4703f900c067ccfd6/pyhomematic/devicetypes/generic.py#L381
10. Attributes to the function are `callback=self._hm_event_callback, bequeath=False, channel=channel` as defined here: https://github.com/home-assistant/home-assistant/blob/2871a650f69257fa6e8e4254527007f964e2f806/homeassistant/components/homematic/__init__.py#L844
11. This causes callbacks with `channel=0` to handled differently than `channel>0`. The former goes into `self._eventcallbacks.append(callback)`. The latter goes into `self._hmchannels[channel]._eventcallbacks.append(callback)`.

I think, this is as far as I can get today. Maybe I'll find more time to trace the callbacks in the coming days... @danielperna84, if this already rings a bell with you, I'll be happy to receive more directions for testing!

**Comment by akloeckner** [comment]
I updated to HA 0.95.4 today. A new sensor appeared for the battery status. I cannot check currently, whether it works. The sabotage attribute is still not working at least (in both sensors: presence and battery).

**Comment by danielperna84** [comment]
The battery sensor entities are not related to the issue with the sabotage attribute.

**Comment by akloeckner** [comment]
This issue is not an issue anymore in the new hahomematic package. 👍 I'll close this issue to get it off my mind. ;-)
